### PR TITLE
ddl: support alter table .. truncate partition (#8624)

### PR DESCRIPTION
**Pull request by tiancaiamao**
Cherry-pick https://github.com/pingcap/tidb/pull/8624 to release-2.1
Need merge parser first https://github.com/pingcap/parser/pull/181

@crazycs520 @winkyao @jackysp

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @zimulala

File: ddl/partition.go [hunk 1]
```diff
@@ -326,8 +338,54 @@ func onDropTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+	if err := job.DecodeArgs(&oldID); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	pi := tblInfo.GetPartitionInfo()
+	if pi == nil {
```
Comment (zimulala) [inline ddl/partition.go]: Do we need to add "job.State = model.JobStateCancelled" here?

File: ddl/partition.go [hunk 2]
```diff
@@ -326,8 +338,54 @@ func onDropTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+			pid, err1 := t.GenGlobalID()
+			if err != nil {
+				return ver, errors.Trace(err1)
+			}
+			def.ID = pid
+			find = true
+			break
+		}
+	}
+	if !find {
```
Comment (zimulala) [inline ddl/partition.go]: Ditto.

File: ddl/partition.go [hunk 3]
```diff
@@ -348,10 +348,12 @@ func onTruncateTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
 	}
 	tblInfo, err := getTableInfo(t, job, job.SchemaID)
 	if err != nil {
+		job.State = model.JobStateCancelled
```
Comment (zimulala) [inline ddl/partition.go]: It is done in `getTableInfo`.

File: ddl/partition.go (hunk 2 above)
Comment (zimulala) [reply ddl/partition.go]: Do we need to do it in master ？

**Comment by tiancaiamao** [comment]
@crazycs520 @winkyao

**Comment by zimulala** [comment]
/run-all-tests tidb-test=release-2.1 tidb-private-test=release-2.1 tikv=release-2.1

**Pull request merged**